### PR TITLE
Create separate benchmark workflow for running large benchmarks

### PR DESCRIPTION
**Pull request by mariecwhite**
Creates a separate workflow for running large benchmarks. Adds ResNet50 at various batch sizes to this workflow. ResNet50 is part of our progression suite that we will be monitoring closely.

Also rearranges the model_groups to be specific to the target hardware.

### CODE REVIEW

**Comment by GMNGeoffrey** [comment]
Whoo thanks for the improvements! I really appreciate you trimming this down

**Comment by GMNGeoffrey** [comment]
Uh, whoops, that comment was intended for https://github.com/openxla/iree/pull/12908. I'm sure this one is great also, but I haven't looked at it yet :-D

File: .github/workflows/benchmark_long.yml [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#
+# Executes long-running benchmarks that are expected to take multple hours.
+
+name: Benchmark Long
+
+on:
+  schedule:
+    - cron: '0 09 * * *'
```
Comment (GMNGeoffrey) [inline .github/workflows/benchmark_long.yml]: Can you add a comment about what this corresponds to? Every day at 09:00 UTC?

File: .github/workflows/benchmark_long.yml [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+  workflow_dispatch:
+  pull_request:
```
Comment (GMNGeoffrey) [inline .github/workflows/benchmark_long.yml]: Don't forget to remove this

File: .github/workflows/benchmark_long.yml [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+          echo "benchmark-presets=cuda-long" >> $GITHUB_OUTPUT
+
+      - name: "Show benchmark presets"
+        env:
+          BENCHMARK_PRESETS: ${{ steps.configure.outputs.benchmark-presets }}
+        run: |
+          echo ":stopwatch: Enabled benchmarks: \`${BENCHMARK_PRESETS}\`" \
+            >> "${GITHUB_STEP_SUMMARY}"
+
+  build_all:
```
Comment (GMNGeoffrey) [inline .github/workflows/benchmark_long.yml:28]: Ah, sad that we have to do this again :-( We should consider indexing things based on commit instead of run ID maybe, although there were reasons I didn't do that. I guess ccache will help.

It would be nice to turn this into a reusable workflow though. There's a fair amount of detail. Could you try and see how that works? My recollection is that passing inputs and outputs around was annoying though

File: .github/workflows/benchmark_long.yml [hunk 4]
```diff
@@ -0,0 +1,499 @@
+# Copyright 2022 The IREE Authors
```
Comment (GMNGeoffrey) [inline .github/workflows/benchmark_long.yml]: This is a lot of copy-paste. Can we extract this into a reusable workflow? https://docs.github.com/en/actions/using-workflows/reusing-workflows

File: build_tools/python/e2e_test_framework/models/model_groups.py [hunk 5]
```diff
@@ -62,4 +31,53 @@
...
+    tf_models.BERT_FOR_MASKED_LM_FP32_SEQLEN512,
+    tf_models.BERT_LARGE_TF_FP32_SEQLEN384,
+    torch_models.MODEL_CLIP_TEXT_SEQLEN64_FP32_TORCH,
+    torch_models.MODEL_UNET_2D_FP32_TORCH,
+    torch_models.EFFICIENTNET_B7_FP32_TORCH,
+]
+
+CUDA_MODELS_LONG = RESNET50_TF_BATCHES
+
+# yapf: disable
```
Comment (GMNGeoffrey) [inline build_tools/python/e2e_test_framework/models/model_groups.py]: Why is this necessary?

File: .github/workflows/benchmark_long.yml (hunk 3 above)
Comment (mariecwhite) [reply .github/workflows/benchmark_long.yml:28]: This is a big refactor. I'll see what I can do.

File: .github/workflows/benchmark_long.yml (hunk 1 above)
Comment (mariecwhite) [reply .github/workflows/benchmark_long.yml]: Added comment and updated to run twice a day.

File: .github/workflows/benchmark_long.yml (hunk 3 above)
Comment (mariecwhite) [reply .github/workflows/benchmark_long.yml:28]: Refactored the common parts to separate workflow files.

File: .github/workflows/benchmark_long.yml (hunk 4 above)
Comment (mariecwhite) [reply .github/workflows/benchmark_long.yml]: Done

File: build_tools/python/e2e_test_framework/models/model_groups.py (hunk 5 above)
Comment (mariecwhite) [reply build_tools/python/e2e_test_framework/models/model_groups.py]: yapf ends up rearranging the list to 2 elements per line instead of one element per line.

Comment (GMNGeoffrey) [reply build_tools/python/e2e_test_framework/models/model_groups.py]: Well that's annoying. I'd think a trailing comma would be taken as intent for one per line. That reminds me that I want to switch us to https://github.com/google/pyink

Comment (ScottTodd) [reply build_tools/python/e2e_test_framework/models/model_groups.py]: You can usually add empty comments after each item in the list to work around that formatter behavior

Comment (GMNGeoffrey) [reply build_tools/python/e2e_test_framework/models/model_groups.py]: That seems worse though :-D

**Review by GMNGeoffrey — APPROVED** [review]
Thanks for the cleanup. Would also like @pzread to weigh in

File: build_tools/python/e2e_test_framework/models/model_groups.py (hunk 5 above)
Comment (ScottTodd) [reply build_tools/python/e2e_test_framework/models/model_groups.py]: ¯\\\_(ツ)\_/¯ eh, I just applied that sort of formatting change in a PR here: https://github.com/openxla/iree/blob/a0d71cb6e2a89942a2bdfb8d037ec1794f47ed61/runtime/bindings/python/tests/vm_types_test.py#L52-L65

File: .github/workflows/benchmark_compilation.yml [hunk 6]
```diff
@@ -0,0 +1,122 @@
+# Copyright 2022 The IREE Authors
```
Comment (pzread) [inline .github/workflows/benchmark_compilation.yml:1]: nit: 2023?

File: build_tools/python/benchmark_suites/iree/module_execution_configs.py [hunk 7]
```diff
@@ -39,6 +39,12 @@ def _with_caching_allocator(
     loader=iree_definitions.RuntimeLoader.NONE,
     driver=iree_definitions.RuntimeDriver.CUDA)
 
+CUDA_CONFIG_LONG = _with_caching_allocator(
```
Comment (pzread) [inline build_tools/python/benchmark_suites/iree/module_execution_configs.py]: It seems like we don't need to create a new execution config for long models. We can directly filter on the model's tags in `export_benchmark_config.py`?

File: build_tools/benchmarks/export_benchmark_config.py [hunk 8]
```diff
@@ -52,7 +52,10 @@
         "x86_64",
     "cuda":
         lambda config: config.target_device_spec.architecture.architecture ==
-        "cuda",
+        "cuda" and "long-running" not in config.module_execution_config.tags,
+    "cuda-long":
+        lambda config: config.target_device_spec.architecture.architecture ==
+        "cuda" and "long-running" in config.module_execution_config.tags,
```
Comment (pzread) [inline build_tools/benchmarks/export_benchmark_config.py]: Instead of filtering on the tag of execution config, maybe we can simply filter on the model's tag?
```suggestion
    "cuda-long":
        lambda config: config.target_device_spec.architecture.architecture ==
        "cuda" and "long-running" in config.module_generation_config.imported_model.model.tags,
```

Comment (mariecwhite) [reply build_tools/benchmarks/export_benchmark_config.py]: We may end up using the same model run under different scenarios so would prefer to not have a 1:1 relationship between a model and how it is executed.

File: .github/workflows/benchmark_compilation.yml (hunk 6 above)
Comment (mariecwhite) [reply .github/workflows/benchmark_compilation.yml:1]: Changed to 2023, including the other yml files.

File: build_tools/python/benchmark_suites/iree/module_execution_configs.py (hunk 7 above)
Comment (mariecwhite) [reply build_tools/python/benchmark_suites/iree/module_execution_configs.py]: See comment below

File: .github/workflows/benchmark_compilation.yml (hunk 6 above)
Comment (GMNGeoffrey) [reply .github/workflows/benchmark_compilation.yml:1]: This isn't a big deal, but we actually *don't* want to update the copyright year on things where the content originally had an older year. So if you adapt some part of a file it should keep 2022 in the copyright. I think that's pretty much the case with all this, though maybe some of the stuff is pretty much entirely new, IDK. Really this whole thing is a silly artifact of how the Apache license is written. I wish it didn't include this requirement for a year :-/

Comment (mariecwhite) [reply .github/workflows/benchmark_compilation.yml:1]: Would all of the new .yml files that I refactored jobs into warrant 2023 since it's a new file (even though it's copy/paste from a 2022 file)?

File: build_tools/benchmarks/export_benchmark_config.py (hunk 8 above)
Comment (pzread) [reply build_tools/benchmarks/export_benchmark_config.py]: To me it sounds like the characteristic of "long benchmark" is the combination of model + execution config, which is represented by `E2EModelRunConfig`:
https://github.com/openxla/iree/blob/4e99afbe7a516e6437f1f4505e3544900e50749f/build_tools/python/e2e_test_framework/definitions/iree_definitions.py#L297-L310

So maybe adding a new "tags" field to `E2EModelRunConfig` is a better way? Another problem to have "long execution config" is that if we want to move benchmarks between "long" and "non-long" groups, we will change their execution configs, which will result in different benchmark series.

File: .github/workflows/benchmark_compilation.yml (hunk 6 above)
Comment (ScottTodd) [reply .github/workflows/benchmark_compilation.yml:1]: https://opensource.google/documentation/reference/copyright#the_year

Comment (mariecwhite) [reply .github/workflows/benchmark_compilation.yml:1]: Thanks for the link Scott. I'll revert to 2022.

File: build_tools/python/benchmark_suites/iree/module_execution_configs.py [hunk 9]
```diff
@@ -39,6 +39,12 @@ def _with_caching_allocator(
...
+    id=unique_ids.IREE_MODULE_EXECUTION_CONFIG_CUDA,
+    tags=["full-inference", "default-flags", "long-running"],
```
Comment (pzread) [inline build_tools/python/benchmark_suites/iree/module_execution_configs.py]: We use the assumption that if "config_a.id = config_b.id", then "config_a" is identical to "config_b". So having two configs with the same id but different tags might run into some issues.

File: build_tools/benchmarks/export_benchmark_config.py (hunk 8 above)
Comment (mariecwhite) [reply build_tools/benchmarks/export_benchmark_config.py]: I see, thanks for pointing that out. I'll add a tags field to E2EModelRunConfig.

File: .github/workflows/benchmark_compilation.yml (hunk 6 above)
Comment (pzread) [reply .github/workflows/benchmark_compilation.yml:1]: Ah, I see. Didn't notice that this is just the code moved around. Sorry for the noise

File: build_tools/benchmarks/export_benchmark_config.py (hunk 8 above)
Comment (mariecwhite) [reply build_tools/benchmarks/export_benchmark_config.py]: Updated to use tags in E2EModelRunConfig. PTAL

File: .github/workflows/benchmark_long.yml [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+    runs-on: ubuntu-20.04
+    env:
+      BENCHMARK_PRESETS: cuda-long
+    outputs:
+      benchmark-presets: ${{ steps.configure.outputs.benchmark-presets }}
+    steps:
+      - name: "Setting benchmark presets"
+        id: configure
+        run: |
+          echo "benchmark-presets=${BENCHMARK_PRESETS}" >> $GITHUB_OUTPUT
```
Comment (pzread) [inline .github/workflows/benchmark_long.yml]: Do we need a step for the benchmark presets? Seems like we just set a constant for it. Maybe define a workflow level env variable and use it below?

```suggestion
env:
    BENCHMARK_PRESETS: cuda-long
```

File: build_tools/python/e2e_test_framework/definitions/iree_definitions.py [hunk 11]
```diff
@@ -324,7 +325,7 @@ def build(cls, module_generation_config: ModuleGenerationConfig,
             module_execution_config: ModuleExecutionConfig,
             target_device_spec: common_definitions.DeviceSpec,
             input_data: common_definitions.ModelInputData,
-            tool: E2EModelRunTool):
+            tool: E2EModelRunTool, tags: Sequence[str]):
```
Comment (pzread) [inline build_tools/python/e2e_test_framework/definitions/iree_definitions.py]: Maybe give it a default value?

```suggestion
            tool: E2EModelRunTool, tags: Optional[Sequence[str]] = None):
    tags = [] if tags is None else list(tags)
```

File: build_tools/python/benchmark_suites/iree/utils.py [hunk 12]
```diff
@@ -15,6 +15,7 @@ def generate_e2e_model_run_configs(
         iree_definitions.ModuleGenerationConfig],
     module_execution_configs: Sequence[iree_definitions.ModuleExecutionConfig],
     device_specs: Sequence[common_definitions.DeviceSpec],
+    tags: Sequence[str] = [],
```
Comment (pzread) [inline build_tools/python/benchmark_suites/iree/utils.py]: It's suggested not to use a mutable object as the initializer. Maybe

```suggestion
    tags: Optional[Sequence[str]] = None,
```

File: build_tools/benchmarks/common/benchmark_suite_test.py [hunk 13]
```diff
@@ -184,7 +184,8 @@ def test_load_from_run_configs(self):
         module_execution_config=exec_config_a,
         target_device_spec=device_spec_a,
         input_data=common_definitions.ZEROS_MODEL_INPUT_DATA,
-        tool=iree_definitions.E2EModelRunTool.IREE_BENCHMARK_MODULE)
+        tool=iree_definitions.E2EModelRunTool.IREE_BENCHMARK_MODULE,
+        tags=[])
```
Comment (pzread) [inline build_tools/benchmarks/common/benchmark_suite_test.py]: No need to specify the empty tags if we give the parameter a default value, see the comment below

File: build_tools/benchmarks/export_benchmark_config_test.py [hunk 14]
```diff
@@ -63,19 +63,22 @@ def test_filter_and_group_run_configs_set_all_filters(self):
         module_execution_config=COMMON_EXEC_CONFIG,
         target_device_spec=device_spec_a,
         input_data=common_definitions.ZEROS_MODEL_INPUT_DATA,
-        tool=iree_definitions.E2EModelRunTool.IREE_BENCHMARK_MODULE)
+        tool=iree_definitions.E2EModelRunTool.IREE_BENCHMARK_MODULE,
+        tags=[])
```
Comment (pzread) [inline build_tools/benchmarks/export_benchmark_config_test.py]: Same here

**Review by pzread — APPROVED** [review]
Nice! Just a few comments

File: build_tools/python/benchmark_suites/iree/utils.py (hunk 12 above)
Comment (GMNGeoffrey) [reply build_tools/python/benchmark_suites/iree/utils.py]: More detail: https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/

File: .github/workflows/benchmark_long.yml (hunk 10 above)
Comment (mariecwhite) [reply .github/workflows/benchmark_long.yml]: Created an env var and removed the step.

File: build_tools/python/e2e_test_framework/definitions/iree_definitions.py (hunk 11 above)
Comment (mariecwhite) [reply build_tools/python/e2e_test_framework/definitions/iree_definitions.py]: Added

File: build_tools/python/benchmark_suites/iree/utils.py (hunk 12 above)
Comment (mariecwhite) [reply build_tools/python/benchmark_suites/iree/utils.py]: Thanks. Added.

File: .github/workflows/benchmark_long.yml (hunk 10 above)
Comment (mariecwhite) [reply .github/workflows/benchmark_long.yml]: This doesn't seem to work. Using an environment variable as a parameter to a re-usable workflow gives an error. It looks like it needs to use an output to a previous job or step: https://github.com/orgs/community/discussions/26671

Comment (GMNGeoffrey) [reply .github/workflows/benchmark_long.yml]: Yeah this is why we have GCS_DIR repeated everywhere. This is a really annoying GitHub limitation. It can just be a literal though and it's only used in one place, so I would just do that.

Comment (mariecwhite) [reply .github/workflows/benchmark_long.yml]: Made it a literal and removed the step.
